### PR TITLE
feat: replace query registry with query serialization

### PR DESCRIPTION
**Pull request by kaisermann**
This PR replaces the O(n) algorithm that checked if two queries were the same with a O(1) version of firestore's own query serialization method: https://github.dev/firebase/firebase-js-sdk/blob/f78ceca1cf9198f5d371320e8814c859c261cf67/packages/firestore/src/core/target.ts#L93-L93

### CODE REVIEW

File: src/modules/Listener.ts [hunk 1]
```diff
@@ -42,11 +42,7 @@ function getFirestoreListenersOnly() {
 }
 
 // istanbul ignore next
-function formatId(id: ListenerId) {
-  if (typeof id === 'string') return `id:${id}`
-
-  return 'query'
```
Comment (gabrielreisn) [inline src/modules/Listener.ts:48]: why did we have it before? fallback to something?

**Review by gabrielreisn — APPROVED** [review]

**Review by klzns — APPROVED** [review]
kind of afraid of firebase changing their internals and breaking us

File: src/modules/Query.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+// Based on
+// https://github.dev/firebase/firebase-js-sdk/blob/master/packages/firestore/src/core/target.ts#L93
+function canonifyQuery(query: Query) {
+  const internal = query._delegate._query
+
+  let canonicalId = internal.path.canonicalString()
+
+  if (internal.collectionGroup != null) {
+    canonicalId += `|cg:${internal.collectionGroup}`
```
[the remark this answers is not in the document]
Comment (kaisermann) [reply src/modules/Query.ts:53]: yep, they are 😳 I'd love to use the internal id if it didn't started as null :(

**Comment by kaisermann** [comment]
@klzns 

Yes, gotta keep track of these changes when updating firestore. I can leave the previous solution as a fallback though

**Comment by klzns** [comment]
@kaisermann not sure how much effort that would be, but that would be good

**Comment by kaisermann** [comment]
@klzns What I can say is that v9 also uses this structure. It's something that we need to be extra careful when updating even a minor version of firestore. 

About fallbacking: maybe wrapping it in a try catch would suffice for most cases

**Comment by kaisermann** [comment]
@klzns however, I think I prefer to know that things broke down instead of relying on the fallback. We shouldn't be updating versions blindly anyway.

**Review by klzns — APPROVED** [review]

**Comment by kaisermann** [comment]
Some benchmarks

image

---

Number of queries: 10 
Query registry: 0.004100993275642395 
Serialization: 0.0012470036745071411 
Better time: serialization 


---

Number of queries: 100 
Query registry: 0.005496993660926819 
Serialization: 0.001216009259223938 
Better time: serialization 



---

Number of queries: 1000 
Query registry: 0.0051580071449279785 
Serialization: 0.0012349933385849 
Better time: serialization 


---

Number of queries: 10000 
Query registry: 0.010857999324798584 
Serialization: 0.001676991581916809 
Better time: serialization 


---
Number of queries: 100000 
Query registry: 0.010164991021156311 
Serialization: 0.00134199857711792 
Better time: serialization 


---

Number of queries: 1000000
Query registry: 0.01437099277973175 
Serialization: 0.0015680044889450073 
Better time: serialization

**Pull request merged**